internal/sockstate/sockstate.go

```diff
@@ -0,0 +1,78 @@
+package sockstate
+
+import (
+	"fmt"
+	"gopkg.in/src-d/go-errors.v1"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type SockState uint8
+
+const (
+	Finished = iota
+	Broken
+	Other
+	Error
+)
+
+var ErrNoSocketLink = errors.NewKind("couldn't resolve file descriptor link to socket")
+
+// GetConnInode returns the Linux inode number of a TCP connection
+func GetConnInode(c *net.TCPConn) (n uint64, err error) {
+	f, err := c.File()
+	if err != nil {
+		return
+	}
+
+	socketStr := fmt.Sprintf("/proc/%d/fd/%d", os.Getpid(), f.Fd())
+	socketLnk, err := os.Readlink(socketStr)
+	if err != nil {
+		return
+	}
+
+	if strings.HasPrefix(socketLnk, sockPrefix) {
+		_, err = fmt.Sscanf(socketLnk, sockPrefix + "%d]", &n)
+		if err != nil {
+			return
+		}
+	} else {
+		err = ErrNoSocketLink.New()
+	}
+	return
+}
+
+// ErrMultipleSocketsForInode is returned when more than one socket is found for an inode
+var ErrMultipleSocketsForInode = errors.NewKind("more than one socket found for inode")
+
+func GetInodeSockState(port int, inode uint64) (SockState, error) {
+	socks, err := tcpSocks(func(s *sockTabEntry) bool {
+		if s.LocalAddr.Port != uint16(port) {
+			return false
+		}
+
+		si, err := strconv.ParseUint(s.Ino, 10, 64)
+		if err != nil {
+			return false
+		}
+		return inode == si
+	})
+	if err != nil {
+		return Error, err
+	}
+
+	switch len(socks) {
+	case 0:
+		return Finished, nil
+	case 1:
+		if socks[0].State == CloseWait {
+			return Broken, nil
+		} else {
+			return Other, nil
```

Comment:
else branch not needed, previous branch already returned

Reply:
True. As a somewhat pedantic note, it's usually considered good practice to add an else branch even when not needed but I also prefer to remove unnecessary branches, lines of code and indentation when possible, so done.

Reply:
In Go that's not considered a good practice. Golang linters complain about it

Reply:
Well, that doesn't suprise me :rofl: